internal/batch/batch_processor.go

```diff
@@ -101,103 +130,249 @@ func (bw *batchWork) estimateSize() int64 {
+	defer close(bp.done)
 	l := log.L(bp.ctx)
-	var batchSize uint
-	var batchPayloadEstimate = batchSizeEstimateBase
-	var lastBatchSealed = time.Now()
-	var quiescing bool
-	var overflowedWork *batchWork
-	for {
-		var timedOut, closed bool
-		if overflowedWork != nil {
-			// We overflowed the size cap when we took this message out the newWork
-			// queue last time round the lop
-			bp.persistWork <- overflowedWork
-			batchSize++
-			batchPayloadEstimate += overflowedWork.estimateSize()
-			overflowedWork = nil
-		} else {
-			// We timeout waiting at the point we think we're ready for disposal,
-			// unless we've started a batch in which case we wait for what's left
-			// of the batch timeout
-			timeToWait := bp.conf.DisposeTimeout
-			if quiescing {
-				timeToWait = 100 * time.Millisecond
-			} else if batchSize > 0 {
-				timeToWait = bp.conf.BatchTimeout - time.Since(lastBatchSealed)
+
+	var batchTimeout = time.NewTimer(bp.conf.DisposeTimeout)
+	idle := true
+	quescing := false
+	for !quescing {
+
+		var timedout, full, overflow bool
+		select {
+		case <-bp.ctx.Done():
+			l.Tracef("Batch processor shutting down")
+			_ = batchTimeout.Stop()
+			return
+		case <-batchTimeout.C:
+			l.Debugf("Batch timer popped")
+			if len(bp.assemblyQueue) == 0 {
+				bp.startQuiesce()
+			} else {
+				// We need to flush
+				timedout = true
 			}
-			timeout := time.NewTimer(timeToWait)
-
-			// Wait for work, the timeout, or close
-			select {
-			case <-timeout.C:
-				timedOut = true
-			case work, ok := <-bp.newWork:
-				if ok && !work.abandoned {
-					workSize := work.estimateSize()
-					if batchSize > 0 && batchPayloadEstimate+workSize > bp.conf.BatchMaxBytes {
-						overflowedWork = work
-					} else {
-						batchSize++
-						batchPayloadEstimate += workSize
-						bp.persistWork <- work
-					}
-				} else {
-					closed = true
+		case work, ok := <-bp.newWork:
+			if !ok {
+				quescing = true
+			} else {
+				full, overflow = bp.addWork(work)
+				if idle {
+					// We've hit a message while we were idle - we now need to wait for the batch to time out.
+					_ = batchTimeout.Stop()
+					batchTimeout = time.NewTimer(bp.conf.BatchTimeout)
+					idle = false
 				}
 			}
-
 		}
+		if (full || timedout || quescing) && len(bp.assemblyQueue) > 0 {
+			// Let Go GC the old timer
+			_ = batchTimeout.Stop()
+
+			// If we are in overflow, start the clock for the next batch to start before we do the flush
+			// (even though we won't check it until after).
+			if overflow {
+				batchTimeout = time.NewTimer(bp.conf.BatchTimeout)
+			}
 
-		// Don't include the sealing time in the duration
-		batchFull := overflowedWork != nil || batchSize >= bp.conf.BatchMaxSize
-		l.Debugf("Assembly batch loop: Size=%d Full=%t Bytes=%.2fkb (est) Overflow=%t", batchSize, batchFull, float64(batchPayloadEstimate)/1024, overflowedWork != nil)
+			err := bp.flush(overflow)
+			if err != nil {
+				l.Tracef("Batch processor shutting down: %s", err)
```

Comment:
Only a trace log here? Should flush errors be more visible?

Reply:
The only way errors can percolate out of `flush` without infinite retry, are if we're shutting down. But will bump to warn.